examples/bert/bert_model.py

```diff
@@ -0,0 +1,709 @@
+                multi-head attention.
+
+        Returns:
+            An output tensor with the same dimensions as input/query tensor.
+        """
+        if isinstance(inputs, (list, tuple)):
+            if len(inputs) == 2:
+                input_tensor, attention_mask = inputs
+                key_value = None
+            elif len(inputs) == 3:
+                input_tensor, key_value, attention_mask = inputs
+            else:
+                raise ValueError(
+                    "Unexpected inputs to %s with length at %d"
+                    % (self.__class__, len(inputs))
+                )
+        else:
+            input_tensor, key_value, attention_mask = (inputs, None, None)
+
+        if self._output_range:
+            if self._norm_first:
+                source_tensor = input_tensor[:, 0 : self._output_range, :]
+                input_tensor = self._attention_layer_norm(input_tensor)
+                if key_value is not None:
+                    key_value = self._attention_layer_norm(key_value)
+            target_tensor = input_tensor[:, 0 : self._output_range, :]
+            if attention_mask is not None:
+                attention_mask = attention_mask[:, 0 : self._output_range, :]
+        else:
+            if self._norm_first:
+                source_tensor = input_tensor
+                input_tensor = self._attention_layer_norm(input_tensor)
+                if key_value is not None:
+                    key_value = self._attention_layer_norm(key_value)
+            target_tensor = input_tensor
+
+        if key_value is None:
+            key_value = input_tensor
+        attention_output = self._attention_layer(
+            query=target_tensor, value=key_value, attention_mask=attention_mask
+        )
+        attention_output = self._attention_dropout(attention_output)
+        if self._norm_first:
+            attention_output = source_tensor + attention_output
+        else:
+            attention_output = self._attention_layer_norm(
+                target_tensor + attention_output
+            )
+        if self._norm_first:
+            source_attention_output = attention_output
+            attention_output = self._output_layer_norm(attention_output)
+        inner_output = self._intermediate_dense(attention_output)
+        inner_output = self._intermediate_activation_layer(inner_output)
+        inner_output = self._inner_dropout_layer(inner_output)
+        layer_output = self._output_dense(inner_output)
+        layer_output = self._hidden_dropout(layer_output)
+
+        if self._norm_first:
+            return source_attention_output + layer_output
+
+        # During mixed precision training, layer norm output is always fp32 for
+        # now. Casts fp32 for the subsequent add.
+        layer_output = tf.cast(layer_output, tf.float32)
+        return self._output_layer_norm(layer_output + attention_output)
+
+
+class PositionEmbedding(keras.layers.Layer):
```

Comment:
Do we ever need position embeddings in a standalone way, or should we make this layer the combination of a regular embedding layer and a position embedding layer? This would be much simpler for end users

Reply:
I am in favor of keeping things modular I think. We might sub out this embedding for a sinusoidal embedding, and might have already performed the term embedding lookup elsewhere.

I think things get less flexible if you fuse, and at least for bert you haven't really the complexity of summing multiple embeddings (still need type embedding for next sentence prediction).

This would be a good discussion to have as we build PositionEmbedding into the library.